app/controllers/two_factor_authentication/add_methods_controller.rb

```diff
@@ -0,0 +1,43 @@
+module TwoFactorAuthentication
+  class AddMethodsController < ApplicationController
+    before_action :confirm_two_factor_authenticated
+
+    def index
+      @two_factor_add_methods_form = TwoFactorAddMethodsForm.new
+      @presenter = two_factor_options_presenter(after_sign_in_path_for(current_user))
+    end
+
+    def create
+      @two_factor_add_methods_form = TwoFactorAddMethodsForm.new
+      result = @two_factor_add_methods_form.submit(two_factor_add_methods_form_params)
+
+      if result.success?
+        process_valid_form
+      else
+        @presenter = two_factor_options_presenter
+        render :index
+      end
+    end
+
+    private
+
+    def two_factor_options_presenter(next_step_url = nil)
+      TwoFactorAddMethodsPresenter.new(current_user, current_sp, view_context, next_step_url)
+    end
+
+    def process_valid_form
+      case @two_factor_add_methods_form.selection
+      when 'sms', 'voice'
+        redirect_to phone_setup_url
+      when 'auth_app'
+        redirect_to authenticator_setup_url
+      when 'piv_cac'
+        redirect_to setup_piv_cac_url
+      end
+    end
+
+    def two_factor_add_methods_form_params
+      params.require(:two_factor_add_methods_form).permit(:selection)
+    end
+  end
+end
```

Comment:
This controller is basically duplicating the existing `TwoFactorAuthenticationSetupController`, which IMO is better named than this one. Why do we need both? Were you planning on only keeping one?

Reply:
The behavior of the setup controller is different from the additional methods screen.

Reply:
The way I see it, they both are presenting the user with a choice of options, with the only difference being that on initial account creation, they automatically get all the options (plus PIV/CAC if eligible), and after account creation, they only see the options that they don't have yet and that they are eligible for.

Is there another difference I'm missing?

Reply:
It's also the context of when it's run and where it's going.  They do have an option to skip this step and we're not fully sure if we will have it automatically come up when we have two backup second factors (ie key and authenticator) or if it will come up until they have all.  Users signing in from SP's won't know new 2fa options are available unless we force this page and allow them to skip it but an argument can be made that it adds more friction.

Reply:
We definitely could make the setup controller more configurable and it would do the job also.  I'm just not a fan of adding all those conditionals.